lib/eslint.js

```diff
@@ -868,10 +870,11 @@ module.exports = (function() {
                         ruleCreator(ruleContext);
 
                     // add all the node types as listeners
-                    Object.keys(rule).forEach(nodeType => {
-                        api.on(nodeType, timing.enabled
-                            ? timing.time(key, rule[nodeType])
-                            : rule[nodeType]
+                    Object.keys(rule).forEach(selector => {
+                        selectors.add(selector);
+                        api.on(selector, timing.enabled
```

Comment:
The approach I was investigating before was to take only the rightmost token of the selector and emit that (mimicking what we already do) and then filter based on the complete query. So you'd add a listener for BlockStatement whose first line checks to see if the query matches before executing. That way, you limit the events emitted while also improving performance by only needing to run a small number of queries for each node rather than re-running every possible query.